File: Project3.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <openssl/aes.h>
#include <openssl/rand.h>
#include <argon2.h>
#include <sqlite3.h>

typedef struct {
    char data[50];
} Node;

typedef struct {
    char request_ip[50];
    char request_timestamp[50];
    int user_id;
} AuthLog;

// Function to encrypt data using AES
void encryptData(const char *data, const char *key, char *encrypted_data) {
    AES_KEY aesKey;
    AES_set_encrypt_key((const unsigned char *)key, 128, &aesKey);
    AES_encrypt((const unsigned char *)data, (unsigned char *)encrypted_data, &aesKey);
}

// Function to decrypt data using AES
void decryptData(const char *encrypted_data, const char *key, char *decrypted_data) {
    AES_KEY aesKey;
    AES_set_decrypt_key((const unsigned char *)key, 128, &aesKey);
    AES_decrypt((const unsigned char *)encrypted_data, (unsigned char *)decrypted_data, &aesKey);
}

// Function to generate a secure password using UUIDv4
void generatePassword(char *password) {
    // Implementation for generating UUIDv4 goes here
    // For simplicity, let's assume it generates a random 16-byte value
    RAND_bytes((unsigned char *)password, 16);
}

// Function to hash the password using Argon2
void hashPassword(const char *password, char *hashed_password) {
    // Configuration for Argon2 hashing
    const uint32_t t_cost = 2;  // Number of iterations
    const uint32_t m_cost = 1 << 16;  // Memory cost
    const uint32_t parallelism = 1;  // Number of threads
    const size_t salt_len = 16;  // Length of the salt
    char salt[salt_len];

    // Generate a random salt
    RAND_bytes((unsigned char *)salt, salt_len);

    // Hash the password using Argon2
    argon2id_hash_raw(t_cost, m_cost, parallelism, password, strlen(password),
                      salt, salt_len, hashed_password, ARGON2_HASH_LEN);
}

// Function to register a new user
void registerUser(sqlite3 *db, const char *username, const char *email) {
    Node user;
    strcpy(user.data, username);

    char key[16];  // 128-bit key for AES
    char encrypted_data[50];
    char decrypted_data[50];
    char password[16];
    char hashed_password[ARGON2_HASH_LEN];

    // Generate AES key
    RAND_bytes((unsigned char *)key, 16);

    // Generate a secure password
    generatePassword(password);

    // Encrypt and store private key
    encryptData(username, key, encrypted_data);

    // Decrypt private key
    decryptData(encrypted_data, key, decrypted_data);

    // Hash the password
    hashPassword(password, hashed_password);

    // Store user details and hashed password in the users table
    char query[256];
    sprintf(query, "INSERT INTO users (username, password_hash, email) VALUES ('%s', '%s', '%s')",
            username, hashed_password, email);
    if (sqlite3_exec(db, query, 0, 0, 0) != SQLITE_OK) {
        fprintf(stderr, "Failed to execute query: %s\n", sqlite3_errmsg(db));
        exit(EXIT_FAILURE);
    }

    // Print the results
    printf("User registered:\n");
    printf("Username: %s\n", username);
    printf("Email: %s\n", email);
    printf("Password: %s\n", password);
}

// Function to log authentication requests
void logAuthenticationRequest(sqlite3 *db, const char *request_ip, const char *request_timestamp, int user_id) {
    AuthLog log;
    strcpy(log.request_ip, request_ip);
    strcpy(log.request_timestamp, request_timestamp);
    log.user_id = user_id;

    // Store authentication log in the auth_logs table
    char query[256];
    sprintf(query, "INSERT INTO auth_logs (request_ip, request_timestamp, user_id) VALUES ('%s', '%s', %d)",
            request_ip, request_timestamp, user_id);
    if (sqlite3_exec(db, query, 0, 0, 0) != SQLITE_OK) {
        fprintf(stderr, "Failed to execute query: %s\n", sqlite3_errmsg(db));
        exit(EXIT_FAILURE);
    }

    // Print the results
    printf("Authentication request logged:\n");
    printf("Request IP: %s\n", request_ip);
    printf("Request Timestamp: %s\n", request_timestamp);
    printf("User ID: %d\n", user_id);
}

int main() {
    // Initialize SQLite database
    sqlite3 *db;
    if (sqlite3_open("your_database.db", &db) != SQLITE_OK) {
        fprintf(stderr, "Cannot open database: %s\n", sqlite3_errmsg(db));
        return EXIT_FAILURE;
    }

    // Example of using the registerUser function
    registerUser(db, "JohnDoe", "john.doe@example.com");

    // Example of using the logAuthenticationRequest function
    logAuthenticationRequest(db, "192.168.1.1", "2023-12-01 12:34:56", 1);

    // Close the database connection
    sqlite3_close(db);

    return 0;
}

File: Project3.py
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend
from argon2 import PasswordHasher
import sqlite3
import uuid
from flask import Flask, request, jsonify

app = Flask(__name__)
DATABASE_FILE = "your_database.db"
AES_KEY = b"your_secret_aes_key"

# Initialize the AES cipher
cipher = Cipher(algorithms.AES(AES_KEY), modes.ECB(), backend=default_backend())
encryptor = cipher.encryptor()
decryptor = cipher.decryptor()

ph = PasswordHasher()

# Database setup
def initialize_database():
    conn = sqlite3.connect(DATABASE_FILE)
    cursor = conn.cursor()

    # Create users table
    cursor.execute('''
        CREATE TABLE IF NOT EXISTS users (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            username TEXT NOT NULL UNIQUE,
            password_hash TEXT NOT NULL,
            email TEXT UNIQUE,
            date_registered TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
            last_login TIMESTAMP
        )
    ''')

    # Create auth_logs table
    cursor.execute('''
        CREATE TABLE IF NOT EXISTS auth_logs (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            request_ip TEXT NOT NULL,
            request_timestamp TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
            user_id INTEGER,
            FOREIGN KEY(user_id) REFERENCES users(id)
        )
    ''')

    conn.commit()
    conn.close()

# Encrypt data using AES
def encrypt_data(data):
    return encryptor.update(data.encode()) + encryptor.finalize()

# Decrypt data using AES
def decrypt_data(data):
    return decryptor.update(data) + decryptor.finalize()

# Generate a secure password using UUIDv4
def generate_password():
    return str(uuid.uuid4())

# Hash the password using Argon2
def hash_password(password):
    return ph.hash(password)

# Register a new user
def register_user(username, email):
    password = generate_password()
    hashed_password = hash_password(password)

    # Encrypt username with AES before storing in the database
    encrypted_username = encrypt_data(username)

    conn = sqlite3.connect(DATABASE_FILE)
    cursor = conn.cursor()

    # Insert user details into the users table
    cursor.execute('''
        INSERT INTO users (username, password_hash, email)
        VALUES (?, ?, ?)
    ''', (encrypted_username, hashed_password, email))

    conn.commit()
    conn.close()

    return password

# Log authentication request
def log_authentication_request(request_ip, user_id):
    conn = sqlite3.connect(DATABASE_FILE)
    cursor = conn.cursor()

    # Insert authentication log into the auth_logs table
    cursor.execute('''
        INSERT INTO auth_logs (request_ip, user_id)
        VALUES (?, ?)
    ''', (request_ip, user_id))

    conn.commit()
    conn.close()

@app.route('/register', methods=['POST'])
def register_endpoint():
    data = request.get_json()
    username = data.get('username')
    email = data.get('email')

    if not username or not email:
        return jsonify({'error': 'Invalid request'}), 400

    try:
        initialize_database()
        password = register_user(username, email)
        return jsonify({'password': password}), 201
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@app.route('/auth', methods=['POST'])
def auth_endpoint():
    data = request.get_json()
    request_ip = request.remote_addr
    username = data.get('username')
    
    if not username:
        return jsonify({'error': 'Invalid request'}), 400

    conn = sqlite3.connect(DATABASE_FILE)
    cursor = conn.cursor()

    # Retrieve user ID from the users table based on the encrypted username
    cursor.execute('''
        SELECT id, username
        FROM users
        WHERE username = ?
    ''', (encrypt_data(username),))

    result = cursor.fetchone()

    if result:
        user_id, encrypted_username = result
        decrypted_username = decrypt_data(encrypted_username).decode()
        log_authentication_request(request_ip, user_id)
        return jsonify({'message': f'Authentication successful for {decrypted_username}'}), 200
    else:
        return jsonify({'error': 'User not found'}), 404

if __name__ == '__main__':
    app.run(debug=True)
